core/state_processor.go

```diff
@@ -49,20 +66,320 @@ func NewStateProcessor(config *params.ChainConfig, bc *BlockChain, engine consen
 	}
 }
 
+type LightStateProcessor struct {
+	randomGenerator *rand.Rand
+	StateProcessor
+}
+
+func NewLightStateProcessor(config *params.ChainConfig, bc *BlockChain, engine consensus.Engine) *LightStateProcessor {
+	randomGenerator := rand.New(rand.NewSource(int64(time.Now().Nanosecond())))
+	return &LightStateProcessor{
+		randomGenerator: randomGenerator,
+		StateProcessor:  *NewStateProcessor(config, bc, engine),
+	}
+}
+
+func (p *LightStateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg vm.Config) (*state.StateDB, types.Receipts, []*types.Log, uint64, error) {
+	allowLightProcess := true
+	if posa, ok := p.engine.(consensus.PoSA); ok {
+		allowLightProcess = posa.AllowLightProcess(p.bc, block.Header())
+		log.Error("===debug, allow to light process?", "allow", allowLightProcess)
+	}
+	// random fallback to full process
+	if check := p.randomGenerator.Int63n(fullProcessCheck); allowLightProcess && check != 0 && len(block.Transactions()) != 0 {
+		var pid string
+		if peer, ok := block.ReceivedFrom.(PeerIDer); ok {
+			pid = peer.ID()
+		}
+		var diffLayer *types.DiffLayer
+		//TODO This is just for debug
+		for tried := 0; tried < diffLayerTimeout; tried++ {
+			// wait a bit for the diff layer
+			diffLayer = p.bc.GetUnTrustedDiffLayer(block.Hash(), pid)
+			if diffLayer != nil {
+				log.Error("===debug find it", "idx", tried)
+				break
+			}
+			time.Sleep(time.Millisecond)
+		}
+		if diffLayer != nil {
+			if err := diffLayer.Receipts.DeriveFields(p.bc.chainConfig, block.Hash(), block.NumberU64(), block.Transactions()); err != nil {
+				log.Error("Failed to derive block receipts fields", "hash", block.Hash(), "number", block.NumberU64(), "err", err)
+				// fallback to full process
+				return p.StateProcessor.Process(block, statedb, cfg)
+			}
+			receipts, logs, gasUsed, err := p.LightProcess(diffLayer, block, statedb, cfg)
+			if err == nil {
+				log.Info("do light process success at block", "num", block.NumberU64())
+				return statedb, receipts, logs, gasUsed, nil
+			} else {
+				log.Error("do light process err at block\n", "num", block.NumberU64(), "err", err)
+				p.bc.removeDiffLayers(diffLayer.DiffHash)
+				// prepare new statedb
+				statedb.StopPrefetcher()
+				parent := p.bc.GetHeader(block.ParentHash(), block.NumberU64()-1)
+				statedb, err = state.New(parent.Root, p.bc.stateCache, p.bc.snaps)
+				if err != nil {
+					return statedb, nil, nil, 0, err
+				}
+				// Enable prefetching to pull in trie node paths while processing transactions
+				statedb.StartPrefetcher("chain")
+			}
+		}
+	}
```

Comment:
do we need to return here or we will always fall back to full process

Reply:
Already fallback to full process.